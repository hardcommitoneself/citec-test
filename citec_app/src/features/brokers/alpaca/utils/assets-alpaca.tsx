// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { AlpacaApi } from "./hooks/api/alpaca-api"

export const assetsAlpaca = async (type: string, tickers: string[]) => {
    
    const getAlpacaAsset = async (type: string, ticker: string) => {
        try {
            const asset = await AlpacaApi({
                type: type,
                url: 'assets/' + ticker,
                method: 'GET',
             })
      
             return asset.data
        } catch (error: any) {
            return error
        }  
    }

    const getAllAssets = async (type: string, tickers: string[]) => {
        
        const assetsPromises: any = []
        tickers.map((ticker) => {
            assetsPromises.push(getAlpacaAsset(type, ticker))
        })

        return await Promise.all(assetsPromises)
    }
    
    return getAllAssets(type, tickers)    
}