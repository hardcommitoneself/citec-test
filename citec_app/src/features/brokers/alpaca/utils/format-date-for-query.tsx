// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
export const formatDateforQuery = (date: Date) => {
    const dformat = [ date.getFullYear(),
        padLeft(date.getMonth()+1),
        padLeft(date.getDate())
        ].join('-');
    
    return dformat
}

function padLeft(n: number){
    return ("00" + n).slice(-2);
  }
 
  
