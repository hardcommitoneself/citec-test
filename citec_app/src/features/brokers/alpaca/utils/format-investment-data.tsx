// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

export const formatInvestmentData = (data: any) => {
    return {'Asset class': data.asset_class,
    'quantity': Number(data.quantity),
    'price': Number(data.price), 
    'Investment': Number(data.investment),  
    'Day gain/loss ($)': Number(data.day_gain_loss[0]),
    'Day gain/loss (%)': Number(data.day_gain_loss[1]),
    'Total gain/loss ($)': Number(data.total_gain_loss[0]),
    'Total gain/loss (%)': Number(data.total_gain_loss[1]),
    }
}
