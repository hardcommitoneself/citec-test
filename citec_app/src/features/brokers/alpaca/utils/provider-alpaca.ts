// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import axios from 'axios'

export const ProviderAlpaca = {
   SetTokenAuth0Alpaca(oauth_code: string) {
      return ProviderAlpaca.setTokenAlpaca(oauth_code)
   },

   async getToken(oauth_code: string) {
      const body = {
         grant_type: 'authorization_code',
         code: oauth_code,
         client_id: process.env.REACT_APP_ALPACA_CLIENT_ID as string,
         client_secret: process.env.REACT_APP_ALPACA_CLIENT_SECRET as string,
         redirect_uri: process.env.REACT_APP_ALPACA_REDIRECT_URI as string,
      }

      const objectBody = Object.keys(body) as Array<keyof typeof body>

      const encodedBody = objectBody
         .map((key) => `${key}=${encodeURIComponent(body[key])}`)
         .join('&')

      let data = {
         code: '',
         value: '',
      }

      await axios({
         headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
         },
         method: 'POST',
         url: 'https://api.alpaca.markets/oauth/token',
         data: encodedBody,
      })
         .then(function (response) {
            data = {
               code: 'sucess',
               value: response.data.access_token,
            }
         })
         .catch(function (error) {
            data = {
               code: 'error',
               value: error,
            }
         })

      return data
   },

   async setTokenAlpaca(code: string) {
      const token = await ProviderAlpaca.getToken(code)
      return token
   },
}
