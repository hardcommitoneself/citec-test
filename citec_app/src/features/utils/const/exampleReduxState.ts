// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

export const exampleReduxState = {
   errors: '',
   taxHarvesting: {
      arrayPositionSelected: [
         {
            outstanding_balance: 0,
            ticker: '$CASH',
            unrealized_capital_gains: 500,
            taxAlpha: -200,
         },
      ],
      positionsTable: [
         {
            outstanding_balance: 50000,
            ticker: 'AAPL',
            unrealized_capital_gains: -1000,
            taxAlpha: 400,
         },
         {
            outstanding_balance: 0,
            ticker: '$CASH',
            unrealized_capital_gains: 500,
            taxAlpha: -200,
         },
      ],
      taxAlpha: -400,
      unrealizedCapitalGains: -500,
      unrealizedCapitalGainsNegative: -1000,
      unrealizedCapitalGainsSelected: 0,
   },
   portfolio: {
      ticker: ['$CASH', 'AAPL'],
      outstanding_balance: [1000000, 50000],
      unrealized_capital_gains: [0, 11234],
      analysis: {
         fundamentals: {
            total_assets: 2,
            total_positions: 2,
            pe: 24.32,
            pb: 38.96,
            dividend_yield: 0.59,
         },
         risk: {
            sharpe: 0.305683,
            volatility: 0.0147,
            beta: 0.048719,
            alpha: 0.000035,
         },
         returns: {
            returns_1M: -0.0021,
            returns_3M: -0.0037,
            returns_6M: -0.0028,
            CAGR: -0.1058,
         },
         capital_gains: {
            outstanding_balance: 1050000,
            unrealized_cap_losses: 0,
            unrealized_cap_gains: 11234,
            realized_cap_losses: 0,
            realized_cap_gains: 0,
            taxes_cost: 0,
            taxes_rate: 0.4,
         },
         sectors: {
            sector: ['Cash', 'Information Technology'],
            percentage: [0.952381, 0.047619],
            percentage_benchmark: [0.3333333333333333, 0.6666666666666666],
         },
         factors: {
            factor: ['style', 'size'],
            percentage: [0.502991, 0.52381],
         },
         segment_matrix: {
            style: ['Val', 'Val'],
            size: ['Large', 'Mid'],
            weight: [0, 0],
         },
         top_holdings: {
            ticker: ['$CASH', 'AAPL'],
            weight: [0.952381, 0.047619],
            outstanding_balance: [1000000, 50000],
            unrealized_capital_gains: [0, 11234],
         },
         concentration: {
            ticker: ['$CASH', 'AAPL'],
            concentration: [95.238095, 100],
         },
         var: {
            quantile: [2, 4],
            value: [0.000072, 0.001334],
         },
         risk_returns: {
            ticker: ['portfolio', 'benchmark'],
            returns: [0.004722, 0.119508],
            volatility: [0.008561, 0.205196],
            sharpe: [0.551568, 0.582407],
         },
      },
   },
   universe: {
      factors_filter: [
         {
            factor: 'style',
            ranges: [0, 100],
         },
         {
            factor: 'size',
            ranges: [0, 100],
         },
         {
            factor: 'yield',
            ranges: [0, 100],
         },
      ],
      sectors_filter: [
         'Industrials',
         'Financials',
         'Health Care',
         'Consumer Staples',
         'Communication Services',
         'Information Technology',
         'Consumer Discretionary',
         'Materials',
         'Cash',
         'Real Estate',
         'Energy',
         'Utilities',
      ],
      to_exclude: [],
      to_include: [],
      universe_base: 'SPDR S&P 500',
   },
   rebalance: {
      ticker: ['$CASH', 'AAPL'],
      hold: [0, 1],
      objective: {
         name: 'Minimize tracking error',
         key: 'min_tracking_error',
      },
      benchmark: 'SPY',
      min_weight: 0.000001,
      number_stocks: [10, 200],
   },
   portfolio_optimized: {
      portfolio_id: null,
      portfolio: 'optimized_portfolio',
      benchmark: 'SPY',
      creation_date: null,
      ticker: ['AAPL', 'MSFT'],
      outstanding_balance: [72506.52, 63477.37],
      unrealized_capital_gains: [0, 0],
      optimization_status: 'Optimization terminated successfully',
   },
   trades: [
      {
         ticker: 'MRK',
         operation: 'SELL',
         volume: 275161.76,
         realized_capital_gains: 0,
         unrealized_capital_gains: 0,
         stocks_number: 3053,
         expected_volume: 275193,
         delta_volume: 31,
         price: 90.1388,
         columnName: {
            url_logo:
               'https://static.finnhub.io/logo/2eedcaa6-80ec-11ea-853f-00000000092a.png',
            name: 'Merck & Co Inc',
         },
         gain_loss: [0, 0],
      },
      {
         ticker: 'BMY',
         operation: 'SELL',
         volume: 152166.26,
         realized_capital_gains: 0,
         unrealized_capital_gains: 0,
         stocks_number: 2072,
         expected_volume: 152146,
         delta_volume: -19,
         price: 73.43,
         columnName: {
            url_logo:
               'https://static.finnhub.io/logo/8b628318-80eb-11ea-93a2-00000000092a.png',
            name: 'Bristol-Myers Squibb Co',
         },
         gain_loss: [0, 0],
      },
      {
         ticker: 'WMT',
         operation: 'SELL',
         volume: 83898.27,
         realized_capital_gains: 0,
         unrealized_capital_gains: 0,
         stocks_number: 679,
         expected_volume: 83845,
         delta_volume: -53,
         price: 123.4832,
         columnName: {
            url_logo:
               'https://static.finnhub.io/logo/52057505dd05a752d364065fcddf7772a700b31c47c39e9dcb41e393f6693d52.png',
            name: 'Walmart Inc',
         },
         gain_loss: [0, 0],
      },
   ],
   workflow: {
      type: 'load',
   },
   static_data: {
      universe: {
         universes: [
            {
               name: 'INVESCO QQQ TRUST SERIES 1',
               description: 'INVESCO QQQ TRUST SERIES 1',
            },
            {
               name: 'ISHARES MSCI USA ESG SELECT',
               description: 'ISHARES MSCI USA ESG SELECT',
            },
         ],
         sectors: [
            'Cash',
            'Communication Services',
            'Consumer Discretionary',
            'Consumer Staples',
         ],
         factors: ['style', 'size', 'yield', 'quality'],
      },
      rebalance: {
         objectives: [
            {
               key: 'min_tracking_error',
               name: 'Minimize tracking error',
            },
            {
               key: 'min_risk_parity',
               name: 'Minimize risk parity',
            },
            { key: 'min_volatility', name: 'Minimize volatility' },
            {
               key: 'max_diversification_ratio',
               name: 'Maximize diversification ratio',
            },
         ],
         benchmarks: [
            {
               key: 'sp_500',
               description: 'S&P 500 universe',
               ticker: 'SPY',
            },
            {
               key: 'nasdaq_100',
               description: 'NASDAQ 100',
               ticker: 'QQQ',
            },
         ],
      },
   },
   brokers: {
      brokerConnected: '',
      balance: 0,
      accountType: [],
      a1c_t: '',
   },
   ideas: [
      {
         portfolio_id: 16,
         portfolio: 'my_first_portfolio_sadasd',
         benchmark: 'SPY',
         creation_date: '2022-04-06 16:10:19',
         ticker: ['MSFT', 'AAPL'],
         outstanding_balance: [200000, 100000],
         unrealized_capital_gains: [0, 35700],
         optimization_status: null,
      },
      {
         portfolio_id: 18,
         portfolio: 'testing',
         benchmark: 'SPY',
         creation_date: '2022-04-06 16:10:19',
         ticker: ['MSFT', 'AAPL'],
         outstanding_balance: [200000, 100000],
         unrealized_capital_gains: [0, 35700],
         optimization_status: null,
      },
      {
         portfolio_id: 19,
         portfolio: 'Jueves',
         benchmark: 'SPY',
         creation_date: '2022-06-23 10:38:00',
         ticker: [
            'CMCSA',
            'REGN',
            'VTR',
            'RTX',
            'NEE',
            'AAPL',
            'MMM',
            'TMO',
            'BLK',
            'MSFT',
            'QCOM',
            'NVDA',
            'CSX',
            'ADBE',
            'XOM',
            'AVGO',
            'BAC',
            'TSLA',
            'SHW',
            'TXN',
            'NFLX',
            'GOOGL',
            'C',
            'MMC',
            'BKNG',
            'COST',
            'NKE',
            'DIS',
            'EL',
            'CSCO',
            'BRK.B',
            'META',
            'AMZN',
            'DG',
            'CCI',
            'ORCL',
            'USB',
            'TGT',
            'PG',
            'ICE',
            'GOOG',
            'AMD',
            'MCD',
            'CVS',
            'HD',
            'CME',
            'MU',
            'UNH',
            'ATVI',
            'V',
            'EOG',
            'GS',
            'HUM',
            'COP',
         ],
         outstanding_balance: [
            80352.37, 74072.31, 59343.2, 54736.44, 41924.05, 41510.68, 35441.43,
            27349.74, 26308.48, 26136.99, 25646.73, 25479.44, 24726.53, 24075.27,
            23856.18, 23763.38, 23735.12, 23092.02, 22995.95, 21888.81, 21775.5,
            21090.31, 16734.94, 16192.87, 15720.33, 14020.04, 14018.61, 13225.5,
            12996.74, 11693.52, 11315.64, 10798.53, 10583.72, 10309.54, 10245.54,
            9611.94, 8747.21, 8723.18, 6339.86, 5866.02, 5426.94, 4912.02, 4857.23,
            3915.03, 3395.13, 3370.41, 3322.34, 3184.94, 3001.78, 1983.75, 1858.2,
            1789.12, 1309.98, 1228.49,
         ],
         unrealized_capital_gains: [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0,
         ],
         optimization_status: null,
      },
      {
         portfolio_id: 20,
         portfolio: 'Min volatility',
         benchmark: 'SPY',
         creation_date: '2022-06-23 11:25:41',
         ticker: [
            'MRK',
            'BMY',
            'WMT',
            'CMCSA',
            'REGN',
            'VTR',
            'RTX',
            'NEE',
            'MMM',
            'CSX',
            'XOM',
            'SHW',
            'TXN',
            'C',
            'BRK.B',
            'BA',
            'BSX',
            'CB',
            'CI',
            'CL',
            'CME',
            'COP',
            'COST',
            'CRM',
            'ZTS',
            'ADBE',
            'AMZN',
            'AVGO',
         ],
         outstanding_balance: [
            275161.76, 152166.26, 83898.27, 80352.37, 74072.31, 59343.2, 54736.44,
            41924.05, 35441.43, 24726.53, 23856.18, 22995.95, 21888.81, 16734.94,
            11315.64, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49,
            2116.49, 2116.49, 2116.49, 119.48, 96.79, 4.71,
         ],
         unrealized_capital_gains: [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0,
         ],
         optimization_status: null,
      },
      {
         portfolio_id: 21,
         portfolio: 'min2',
         benchmark: 'SPY',
         creation_date: '2022-06-23 11:27:32',
         ticker: [
            'MRK',
            'BMY',
            'WMT',
            'CMCSA',
            'REGN',
            'VTR',
            'RTX',
            'NEE',
            'MMM',
            'CSX',
            'XOM',
            'SHW',
            'TXN',
            'C',
            'BRK.B',
            'BA',
            'BSX',
            'CB',
            'CI',
            'CL',
            'CME',
            'COP',
            'COST',
            'CRM',
            'ZTS',
            'ADBE',
            'AMZN',
            'AVGO',
         ],
         outstanding_balance: [
            275161.76, 152166.26, 83898.27, 80352.37, 74072.31, 59343.2, 54736.44,
            41924.05, 35441.43, 24726.53, 23856.18, 22995.95, 21888.81, 16734.94,
            11315.64, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49, 2116.49,
            2116.49, 2116.49, 2116.49, 119.48, 96.79, 4.71,
         ],
         unrealized_capital_gains: [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0,
         ],
         optimization_status: null,
      },
   ],
}
