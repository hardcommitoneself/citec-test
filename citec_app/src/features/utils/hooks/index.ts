// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

export * from './use-app-navigate'
export * from './use-app-redux-loader'
export * from './use-app-subscriptions'
export * from './use-debounce'
export * from './use-delete-idea'
export * from './use-fetch-api'
export * from './use-form'
export * from './use-get-auth0-token'
export * from './use-get-custom-universe'
export * from './use-get-optimization'
export * from './use-get-summary'
export * from './use-handler-tax-harvesting'
export * from './use-select-universe'
export * from './use-validate-criteria-selected'
export * from './use-validate-number-stocks'
export * from './use-validate-portfolio-loaded'
export * from './use-validate-registered-user'
export * from './use-validate-sold-position'
export * from './use-validate-ticker'
export * from './use-validate-tickers'
