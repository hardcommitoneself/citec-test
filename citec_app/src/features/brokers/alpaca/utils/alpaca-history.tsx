// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.

import { AlpacaApi } from 'features/brokers/alpaca/utils/hooks/api/alpaca-api'
import { formatDateforQuery } from 'features/brokers/alpaca/utils/format-date-for-query'
import { formatHistoryData } from 'features/brokers/alpaca/utils/format-history-data'

export const AlpacaHistory = (
   typeAccount: string,
   typeHistory: string,
   callback: any,
   callbackError: any
) => {
   const today = new Date()

   const params = {
      timeframe: '1D',
      date_end: formatDateforQuery(today),
      extended_hours: true,
      period: '1D',
      date_start: '',
   }

   switch (typeHistory) {
      case '1D': {
         params.timeframe = '5Min'
         params.period = '1D'
         break
      }
      case '1W': {
         params.timeframe = '1D'
         params.period = '1W'
         break
      }
      case '1M': {
         params.period = '1M'
         break
      }

      case '3M': {
         params.period = '3M'
         break
      }
      case '1Y': {
         params.period = '1A'
         break
      }
      case 'YTD': {
         params.period = '1A'
         params.date_end = ''
         params.date_start = formatDateforQuery(new Date(today.getFullYear(), 0, 1))
         break
      }

      default:
   }

   AlpacaApi({
      type: typeAccount,
      url: 'account/portfolio/history',
      method: 'GET',
      params: params,
   })
      .then(function (response) {
         if (response.data.equity.length == 0) {
            setDataHistory([], 0, 0, callback)
         } else {
            const actualEquity = getActualEquity(response.data.equity)
            const firstEquity = getFirstEquityValid(response.data.equity)

            const equityFinal = getFinalEquityWhitouthZero(
               response.data.equity,
               firstEquity as number
            )

            const finalTimestamp = getFinalTimestampMiliseconds(
               response.data.timestamp,
               typeHistory
            )
            const datos = formatHistoryData(finalTimestamp, equityFinal)

            if (typeHistory == '1D') {
               const params = {
                  timeframe: '1D',
                  date_end: formatDateforQuery(today),
                  extended_hours: true,
                  period: '1W',
               }

               AlpacaApi({
                  type: typeAccount,
                  url: 'account/portfolio/history',
                  method: 'GET',
                  params: params,
               })
                  .then(function (res) {
                     const lastPosition = res.data.equity.length - 2
                     const firstEquity = res.data.equity[lastPosition]
                     setDataHistory(datos, actualEquity, firstEquity, callback)
                  })
                  .catch(function (err) {
                     callbackError(err)
                  })
            } else {
               setDataHistory(datos, actualEquity, firstEquity, callback)
            }
         }
      })
      .catch(function (error) {
         callbackError(error)
      })
}

export function getFinalEquityWhitouthZero(
   arrayEquity: number[],
   value: number
): number[] {
   const equityFinal = arrayEquity.map(function (eq: number) {
      return eq == 0 ? value : eq
   })

   return equityFinal
}

export function getFinalTimestampMiliseconds(
   arrayTimestamp: number[],
   typeH: string
): number[] {
   if (typeH == '1D') return arrayTimestamp

   const dayInSeconds = 60 * 60 * 24 * 1
   const finalTimestamp = arrayTimestamp.map(function (timestmp, index) {
      return index != arrayTimestamp.length - 1 ? timestmp - dayInSeconds : timestmp
   })

   return finalTimestamp
}

export function getActualEquity(arrayEquity: number[]) {
   const filteredEquitys = arrayEquity.filter((equity: number) => equity !== null)
   const last = filteredEquitys.length - 1

   return filteredEquitys[last]
}

export function getFirstEquityValid(arrayEquity: number[]): number {
   const firstNumber = arrayEquity.find((element: number) => {
      return element != 0
   })

   return firstNumber as number
}

function setDataHistory(
   datos: number[][],
   ActualEquity: number,
   firstEquity: number,
   callback: any
) {
   const profitLoss = ActualEquity - (firstEquity as number)
   const profitLossPer =
      ((ActualEquity - (firstEquity as number)) / (firstEquity as number)) * 100

   callback(datos, ActualEquity, profitLoss, profitLossPer)
}
