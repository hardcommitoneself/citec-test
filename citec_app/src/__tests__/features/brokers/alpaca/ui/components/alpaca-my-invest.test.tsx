// Copyright (C) 2021-Present CITEC Inc. <https://citecsolutions.com/>
// All rights reserved
//
// This file is part of CITEC Inc. source code.
// This software framework contains the confidential and proprietary information
// of CITEC Inc., its affiliates, and its licensors. Your use of these
// materials is governed by the terms of the Agreement between your organisation
// and CITEC Inc., and any unauthorised use is forbidden. Except as otherwise
// stated in the Agreement, this software framework is for your internal use
// only and may only be shared outside your organisation with the prior written
// permission of CITEC Inc.
// CITEC Inc. source code can not be copied and/or distributed without the express
// permission of CITEC Inc.
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import {
   AlpacaMyInvest,
   handleOnExportInvestments,
   loadPortfolioOnRedux
} from 'features/brokers/alpaca/ui/components/alpaca-my-invest'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { BrowserRouter as Router, MemoryRouter } from 'react-router-dom'
import { renderHook } from '@testing-library/react-hooks'
import { exampleReduxState } from 'features/utils/const/exampleReduxState'
import { ReactNode } from 'react'
import { act } from 'react-dom/test-utils'
import { useAppNavigate, useFetchApi } from 'features/utils/hooks'

process.env.REACT_APP_ALPACA_VISIBILITY = 'true'

const mockUseFetchApi = useFetchApi as jest.MockedFunction<typeof useFetchApi>
jest.mock('features/utils/hooks/use-fetch-api')

const mockUseAppNavigate = useAppNavigate as jest.MockedFunction<
   typeof useAppNavigate
>
jest.mock('features/utils/hooks/use-app-navigate')

const mockedUseDispatch = jest.fn()

jest.mock('react-redux', () => ({
   ...jest.requireActual('react-redux'),
   useDispatch: () => mockedUseDispatch,
}))

describe('My investments tests', () => {

   const mockedReps = [
      {
         ticker: 'VZ',
         status: 'OK',
         isin: 'US92343V1044',
         figi: 'BBG000HS77T5',
         name: 'Verizon Communications Inc',
         ipo: '1984-02-15',
         sector: 'Communication Services',
         url_morningstar: 'https://www.morningstar.com/stocks/xnys/VZ/quote',
         url_company: 'https://www.verizon.com/',
         url_logo:
            'https://static.finnhub.io/logo/9b7ee55a-80ec-11ea-bc1f-00000000092a.png',
         outstanding_balance: 0,
         unrealized_capital_gains: 0,
      },
   ]

   const mockedIdea = [
      {
         id: '0.0',
         original: {
            portfolio_id: 105,
            portfolio: 'VZ',
            benchmark: 'SPY',
            investment: 23369.4,
            unrealized_capital_gains: 0,
            creation_date: '2022-07-27 15:56:20',
            columnName: {
               ticker: 'VZ',
               status: 'OK',
               isin: 'US92343V1044',
               figi: 'BBG000HS77T5',
               name: 'Verizon Communications Inc',
               ipo: '1984-02-15',
               sector: 'Communication Services',
               url_morningstar: 'https://www.morningstar.com/stocks/xnys/VZ/quote',
               url_company: 'https://www.verizon.com/',
               url_logo:
                  'https://static.finnhub.io/logo/9b7ee55a-80ec-11ea-bc1f-00000000092a.png',
               outstanding_balance: 0,
               unrealized_capital_gains: 0,
            },
         },
         index: 0,
         depth: 1,
         cells: [{}],
         values: {
            investment: 23369.4,
            total_gain_loss: [0, 0]
         },
         originalSubRows: [],
         subRows: [],
      },
   ]

   beforeEach(() => {
      mockUseFetchApi.mockReturnValue({
         resp: mockedReps,
         isLoading: false,
         error: '',
      })

      mockUseAppNavigate.mockReturnValue(jest.fn())
   })

   const mockStore = configureStore()


   test('renders My investments', () => {

      const initialState = {
         brokers: {
            brokerConnected: 'Alpaca',
            cash: 0,
            accountType: ['PAPER'],
            a1c_t: '',
            accountTypeSelect: 'PAPER',
            buyingPower: 10000,
         },
      }
      
      const store = mockStore(initialState)

      const { container } = render(
         <Router>
            <Provider store={store}>
               <AlpacaMyInvest />
            </Provider>
         </Router>
      )
      const linkElement = screen.getByText(/My Investments/i)
      expect(linkElement).toBeInTheDocument()
      expect(container).toMatchSnapshot()
   })

   test('loadPortfolioOnRedux', () => {

      const store = mockStore(exampleReduxState)

      const wrapper = ({ children }: { children: ReactNode }) => (
         <MemoryRouter>
            <Provider store={store}>{children}</Provider>
         </MemoryRouter>
      )

      const { result } = renderHook(() => AlpacaMyInvest(), { wrapper })

      act(() => {
         loadPortfolioOnRedux(mockedIdea)
      })

      expect(mockUseAppNavigate).toBeCalled()
      expect(mockUseAppNavigate).toBeCalledTimes(1)
      expect(mockedUseDispatch).toBeCalled()
      expect(mockedUseDispatch).toBeCalledTimes(4)
   })

   test("Test handleOnExportInvestments valid Data", () => {

      const data = {
         values: {
            asset_class: "ETF",
            day_gain_loss: [3.2, 0.51],
            expander: undefined,
            investment: 630.62,
            options: undefined,
            price: 0,
            quantity: 2,
            total_gain_loss: [0.48, 0.08],
         },
         subRows: [
            {
               values: {
                  asset_class: "PSA",
                  day_gain_loss: ['3.2', 0.51],
                  expander: undefined,
                  investment: "630.62",
                  options: undefined,
                  price: "315.31",
                  quantity: "2",
                  total_gain_loss: ['0.48', 0.08],
               }
            },
            {
               values: {
                  asset_class: "GE",
                  day_gain_loss: ['-1.43', -2.12],
                  expander: undefined,
                  investment: "66.08",
                  options: undefined,
                  price: "66.08",
                  quantity: "1",
                  total_gain_loss: ['-1.43', -2.12],
               }
            },
         ]

      }

      expect(handleOnExportInvestments(data)).toEqual(1)

   })

   test("Test handleOnExportInvestments invalid Data", () => {

      expect(handleOnExportInvestments([])).not.toEqual(1)

   })

})