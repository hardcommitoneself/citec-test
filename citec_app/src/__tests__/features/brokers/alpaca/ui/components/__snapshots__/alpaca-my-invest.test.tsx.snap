// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My investments tests renders My investments 1`] = `
<div>
  <div
    class="mx-10 mt-8 bg-white p-4 rounded-2xl relative balance"
  >
    <div
      class="flex buying-power"
    >
      <h3
        class="font-semibold text-[26px] text-[#2d405a]"
      >
        My Investments
      </h3>
    </div>
    <div
      class="p-4 flex justify-center items-center"
    >
      <span
        class="text-lg text-neutral-500 font-semibold"
      />
    </div>
  </div>
</div>
`;
